pkg/server/plugin/upstreamauthority/spire/spire.go

```diff
@@ -70,59 +88,141 @@ func (m *Plugin) GetPluginInfo(context.Context, *plugin.GetPluginInfoRequest) (*
 	return &plugin.GetPluginInfoResponse{}, nil
 }
 
+func (m *Plugin) SetLogger(log hclog.Logger) {
+	m.pollMtx.Lock()
+	defer m.pollMtx.Unlock()
+	m.log = log.Named("upstream-authority-spire")
+}
+
 func (m *Plugin) MintX509CA(request *upstreamauthority.MintX509CARequest, stream upstreamauthority.UpstreamAuthority_MintX509CAServer) error {
-	m.mtx.RLock()
-	defer m.mtx.RUnlock()
-	ctx := stream.Context()
+	m.subscribeToPolling()
+	defer m.unsubscribeToPolling()
 
-	wCert, wKey, wBundle, err := m.getWorkloadSVID(ctx, m.config)
+	m.nodeMtx.RLock()
+	certChain, _, err := submitCSRUpstreamCA(stream.Context(), m.nodeClient, request.Csr)
+	m.nodeMtx.RUnlock()
 	if err != nil {
 		return err
 	}
 
-	conn, err := m.newNodeClientConn(ctx, wCert, wKey, wBundle)
-	if err != nil {
-		return err
+	ticker := time.NewTicker(internalPollFreq)
+	var rootCAs []*common.Certificate
+	for {
+		select {
+		case <-ticker.C:
+			newRootCAs := m.getBundle().RootCas
+			if !areRootsEqual(rootCAs, newRootCAs) {
+				rootCAs = newRootCAs
+				err := stream.Send(&upstreamauthority.MintX509CAResponse{
+					X509CaChain:       certsToRawCerts(certChain),
+					UpstreamX509Roots: commonCertsToRawCerts(rootCAs),
+				})
+				if err != nil {
+					m.log.Error("cannot send X.509 CA chain and roots", "error", err)
+				}
+			}
+
+		case <-stream.Context().Done():
+			ticker.Stop()
+			return nil
+		}
 	}
-	nodeClient := node.NewNodeClient(conn)
-	defer conn.Close()
+}
 
-	certChain, bundle, err := m.submitCSRUpstreamCA(ctx, nodeClient, request.Csr)
+func (m *Plugin) PublishJWTKey(req *upstreamauthority.PublishJWTKeyRequest, stream upstreamauthority.UpstreamAuthority_PublishJWTKeyServer) error {
+	m.subscribeToPolling()
+	defer m.unsubscribeToPolling()
+
+	m.nodeMtx.RLock()
+	_, err := m.nodeClient.PushJWTKeyUpstream(stream.Context(), &node.PushJWTKeyUpstreamRequest{JwtKey: req.JwtKey})
+	m.nodeMtx.RUnlock()
 	if err != nil {
 		return err
 	}
 
-	return stream.Send(&upstreamauthority.MintX509CAResponse{
-		X509CaChain:       certsToRawCerts(certChain),
-		UpstreamX509Roots: certsToRawCerts(bundle.RootCAs()),
-	})
+	ticker := time.NewTicker(internalPollFreq)
+	var keys []*common.PublicKey
+	for {
+		select {
+		case <-ticker.C:
+			newKeys := m.getBundle().JwtSigningKeys
+			if !arePublicKeysEqual(keys, newKeys) {
```

Comment:
We have to be careful here. The first time through the keys could compare equal since the bundle may not have been fetched yet (both will be empty). If we're relying on the keys not being equal to send the first response, this may not do that.

More critically, we have to make sure the keys here is as fresh or fresher than the keys returned by the `PushJWTKeyUpstream` call.

Reply:
Good point! I'll update `m.currentBundle` with the keys contained in the `PushJWTKeyUpstream` response.